### PR TITLE
py-aquapark-v1

### PR DESCRIPTION
**Pull request by muriakula**

### CODE REVIEW

File: app/main.py [hunk 1]
```diff
@@ -2,24 +2,70 @@
 
 
 class IntegerRange:
-    pass
+    def __init__(self, min_amount: int, max_amount: int) -> None:
+        self.min_amount = min_amount
+        self.max_amount = max_amount
+
+    def __set_name__(self, owner: type, name: str) -> None:
```
Comment (olena-hul) [inline app/main.py]: Wrong annotation for owner

File: app/main.py [hunk 2]
```diff
@@ -2,24 +2,70 @@
...
+        self.protected_name = "_" + name
+
+    def __get__(self, instance: object, owner: type) -> int:
+        return getattr(instance, self.protected_name)
+
+    def __set__(self, instance: object, value: int) -> None:
```
Comment (olena-hul) [inline app/main.py]: Wrong annotation for instance and owner

File: app/main.py [hunk 3]
```diff
@@ -2,24 +2,70 @@
...
+        if isinstance(value, int):
+            if self.min_amount <= value <= self.max_amount:
+                setattr(instance, self.protected_name, value)
```
Comment (olena-hul) [inline app/main.py:31]: raise ValueError here, when the condition is not true. Then you can catch this ValueError inside can_access method of Slide

File: app/main.py [hunk 4]
```diff
@@ -2,24 +2,70 @@
...
+        self.age_range = age_range
+        self.height_range = height_range
+        self.weight_range = weight_range
+
+    def validate(self, visitor: Visitor) -> bool:
+        return (
+            self.age_range[0] <= visitor.age <= self.age_range[1]
+            and self.weight_range[0] <= visitor.weight <= self.weight_range[1]
+            and self.height_range[0] <= visitor.height <= self.height_range[1]
+        )
```
Comment (olena-hul) [inline app/main.py]: You won't need this method when the ValueError is risen

**Review by olena-hul — CHANGES_REQUESTED** [review]

File: app/main.py (hunk 2 above)
Comment (muriakula) [reply app/main.py]: What annotation should owner and instanse have?

File: app/main.py [hunk 5]
```diff
@@ -1,25 +1,79 @@
...
+from typing import Type
 
 
 class IntegerRange:
-    pass
+    def __init__(self, min_amount: int, max_amount: int) -> None:
+        self.min_amount = min_amount
+        self.max_amount = max_amount
+
+    def __set_name__(self, owner: Type, name: str) -> None:
```
Comment (Bohdan-Salamakha) [inline app/main.py]: incorrect "owner" typehint

File: app/main.py [hunk 6]
```diff
@@ -1,25 +1,79 @@
...
+        self.protected_name = "_" + name
+
+    def __get__(self, instance: Type, owner: Type) -> int:
```
Comment (Bohdan-Salamakha) [inline app/main.py]: incorrect "owner", "instance" typehint

File: app/main.py [hunk 7]
```diff
@@ -1,25 +1,79 @@
...
+        return getattr(instance, self.protected_name)
+
+    def __set__(self, instance: Type, value: int) -> None:
```
Comment (Bohdan-Salamakha) [inline app/main.py]: incorrect "instance" typehint

File: app/main.py [hunk 8]
```diff
@@ -1,25 +1,79 @@
...
+        if isinstance(value, int):
+            if self.min_amount <= value <= self.max_amount:
+                setattr(instance, self.protected_name, value)
+                return
+            else:
```
Comment (Bohdan-Salamakha) [inline app/main.py]: remove redundant "else"

**Review by Bohdan-Salamakha — CHANGES_REQUESTED** [review]
v1

File: app/main.py [hunk 9]
```diff
@@ -1,25 +1,90 @@
...
+    def __set__(
+            self,
+            instance: Union["SlideLimitationValidator", "Visitor"],
+            value: int
+    ) -> None:
+        if isinstance(value, int):
+            if self.min_amount <= value <= self.max_amount:
+                setattr(instance, self.protected_name, value)
+                return
+            raise ValueError
```
Comment (ArsenPidhoretskyi) [inline app/main.py]:
```suggestion
       raise ValueError
```
Wrong tabulation. Also add description for exception

File: app/main.py [hunk 10]
```diff
@@ -1,25 +1,90 @@
...
+            self,
+            owner: Union["SlideLimitationValidator", "Visitor"],
+            name: str
+    ) -> None:
+        self.protected_name = "_" + name
+
+    def __get__(
+            self,
+            instance: Union["SlideLimitationValidator", "Visitor"],
+            owner: Union["SlideLimitationValidator", "Visitor"]
```
Comment (ArsenPidhoretskyi) [inline app/main.py]: owner is type
Use debuging and just do
print(type(instance))
print(type(owner))

File: app/main.py [hunk 11]
```diff
@@ -1,25 +1,90 @@
...
+                setattr(instance, self.protected_name, value)
+                return
+            raise ValueError
 
 
 class Visitor:
-   pass
+    age = IntegerRange(4, 60)
+    height = IntegerRange(80, 220)
+    weight = IntegerRange(20, 120)
```
Comment (ArsenPidhoretskyi) [inline app/main.py]: This is not needed here

File: app/main.py [hunk 12]
```diff
@@ -1,25 +1,90 @@
...
 
 
 class ChildrenSlideLimitationValidator(SlideLimitationValidator):
-    pass
+    age = IntegerRange(4, 14)
+    height = IntegerRange(80, 120)
+    weight = IntegerRange(20, 50)
+
+    def __init__(self, age: int, height: int, weight: int) -> None:
+        super().__init__(age, height, weight)
```
Comment (ArsenPidhoretskyi) [inline app/main.py]: This is works by default. Redundant code remove it the same inside AdultSlideLimitationValidator

**Review by ArsenPidhoretskyi — CHANGES_REQUESTED** [review]

File: app/main.py [hunk 13]
```diff
@@ -9,42 +9,38 @@ def __init__(self, min_amount: int, max_amount: int) -> None:
 
     def __set_name__(
             self,
-            owner: Union["SlideLimitationValidator", "Visitor"],
+            owner: ABCMeta,
```
Comment (ArsenPidhoretskyi) [inline app/main.py]: type[SlideLimitationValidator]
And also in other places

File: app/main.py [hunk 14]
```diff
@@ -9,42 +9,38 @@ def __init__(self, min_amount: int, max_amount: int) -> None:
...
-            owner: Union["SlideLimitationValidator", "Visitor"]
+            instance: "SlideLimitationValidator",
+            owner: ABCMeta
     ) -> int:
         return getattr(instance, self.protected_name)
 
     def __set__(
             self,
-            instance: Union["SlideLimitationValidator", "Visitor"],
+            instance: "AdultSlideLimitationValidator",
```
Comment (ArsenPidhoretskyi) [inline app/main.py]: instance here will be the same as in the get: SlideLimitationValidator

**Review by ArsenPidhoretskyi — CHANGES_REQUESTED** [review]
A little changes

**Review by ArsenPidhoretskyi — APPROVED** [review]
Well done!

**Pull request closed without merging**